### PR TITLE
Implemented missing events

### PR DESCRIPTION
**Pull request by DigitalChiller**
I implemented the missing events, and made some minor improvements.

### CODE REVIEW

File: src/types/content.rs [hunk 1]
```diff
@@ -0,0 +1,176 @@
...
+/// Wether guests can join this room
+pub enum GuestAccess {
+    CanJoin,
+    Forbidden,
+}
+
+#[derive(Serialize, Deserialize, Debug)]
+pub struct Receipts {
+    #[serde(rename="m.read")]
+    read: HashMap<String,Receipt>
```
Comment (eeeeeta) [inline src/types/content.rs]: fields in this file should be marked `pub`, so users can actually get at them (it's easy to forget :)

File: src/types/messages.rs [hunk 2]
```diff
@@ -0,0 +1,242 @@
...
+    #[serde(rename="m.call.hangup")]
+    /// Sent by either party to signal their termination of the call. This can be sent either once the call has has been established or before to abort the call.
+    CallHangup {
+        /// Required. The ID of the call this event relates to.
+        call_id: String,
+        /// Required. The version of the VoIP specification this message adheres to. This specification is version 0.
+        version: i32,
+    },
+}
+
```
Comment (eeeeeta) [inline src/types/messages.rs]: seems to be a lot of empty space here for no reason...

File: Cargo.toml [hunk 3]
```diff
@@ -19,3 +19,8 @@ serde = "0.9"
 serde_derive = "0.9"
 serde_json = "0.9"
 tokio-core = "0.1"
+rpassword = "0.4.2"
```
Comment (eeeeeta) [inline Cargo.toml]: List this under `[dev-dependencies]` instead (it's a dependent of an example, but is not used in the main crate; c.f. http://doc.crates.io/specifying-dependencies.html)

Comment (eeeeeta) [reply Cargo.toml]: Also, bump the `version` to `0.6.0`, and feel free to add yourself under the `authors` list :)

File: Cargo.toml [hunk 4]
```diff
@@ -19,3 +19,8 @@ serde = "0.9"
...
+
+[features]
+gitm_deny_unknown = []
```
Comment (eeeeeta) [inline Cargo.toml]: bikeshed: are the `gitm_` prefixes really necessary?

File: examples/echo.rs [hunk 5]
```diff
@@ -1,26 +1,33 @@
 extern crate glitch_in_the_matrix as gm;
 extern crate futures;
 extern crate tokio_core;
+extern crate rpassword;
 
 use futures::{Future, Stream};
 use tokio_core::reactor::Core;
 use gm::{MatrixClient, MatrixFuture};
-use gm::types::{Content, Message};
+// use gm::types::{EventTypes, Content, Message};
```
Comment (eeeeeta) [inline examples/echo.rs]: this line can probably be deleted (?)

File: examples/echo.rs [hunk 6]
```diff
@@ -1,26 +1,33 @@
...
-    println!("server: {}\nusername: {}\npassword: {}", server, username, password);
+    let (server, username) = (&args[0], &args[1]);
+    println!("Type password for the bot (characters won't show up as you type them)");
+    let password = &prompt_password_stdout("password:").unwrap();
+    let mut core = Core::new().unwrap();
+    let hdl = core.handle();
     let mut mx = core.run(MatrixClient::login(username, password, server, &hdl)).unwrap();
-    println!("[+] Connected to Matrix.");
-    let mut ss = mx.get_sync_stream();
+    println!("[+] Connected to {} as {}",server,username);
```
Comment (eeeeeta) [inline examples/echo.rs]: nit: ideally format this with spaces like `println!("[+] Connected to {} as {}", server, username);`

File: src/lib.rs [hunk 7]
```diff
@@ -284,14 +288,24 @@ impl MatrixClient {
     /// Wrapper function that sends a `Message::Notice` with the specified unformatted text
     /// to the given room ID. Provided for convenience purposes.
     pub fn send_simple<T: Into<String>>(&mut self, roomid: &str, msg: T) -> MatrixFuture<SendReply> {
-        let msg = Message::Notice { body: msg.into(), formatted_body: None, format: None };
+        let msg = Message::Notice{
```
Comment (eeeeeta) [inline src/lib.rs]: nit: ideally format this with a space like `Message::Notice {`

File: src/lib.rs [hunk 8]
```diff
@@ -109,6 +111,8 @@ impl<T: Deserialize> Future for ResponseWrapper<T> {
 
     fn poll(&mut self) -> Poll<Self::Item, Self::Error> {
         let resp = try_ready!(self._poll());
+        #[cfg(feature="gitm_show_responses")]
+        println!("{:#}",String::from_utf8(resp.to_vec()).unwrap());
```
Comment (eeeeeta) [inline src/lib.rs]: nit: ideally format this with a space like `println!("{:#}", String::from_utf8(resp.to_vec()).unwrap());`

File: src/lib.rs [hunk 9]
```diff
@@ -284,14 +288,24 @@ impl MatrixClient {
...
+            format: None
+        };
         self.send(roomid, msg)
     }
     /// Wrapper function that sends a `Message::Notice` with the specified HTML-formatted text
     /// (and accompanying unformatted text, if given) to the given room ID.
     pub fn send_html<T: Into<String>, U: Into<Option<String>>>(&mut self, roomid: &str, msg: T, unformatted: U) -> MatrixFuture<SendReply> {
         let m = msg.into();
-        let msg = Message::Notice { body: unformatted.into().unwrap_or(m.clone()), formatted_body: Some(m), format: Some("org.matrix.custom.html".into()) };
+        let msg = Message::Notice{
```
Comment (eeeeeta) [inline src/lib.rs]: similarly here

File: src/types/content.rs [hunk 10]
```diff
@@ -0,0 +1,176 @@
...
+    Join,
+    Leave,
+    Ban,
+    // reserved word
+    // Knock,
+}
+
+#[derive(Serialize, Deserialize, Debug)]
+#[serde(rename_all="snake_case")]
+/// Defines what Membership a user in a room
```
Comment (eeeeeta) [inline src/types/content.rs]: I think you got a bit copy-pasta happy here... either remove, or replace with something that makes sense :P

File: src/types/content.rs [hunk 11]
```diff
@@ -0,0 +1,176 @@
...
+pub enum HistoryVisibility {
+    Invited,
+    Joined,
+    Shared,
+    WorldReadable,
+}
+#[derive(Serialize, Deserialize, Debug)]
+#[serde(rename_all="snake_case")]
+/// Wether guests can join this room
```
Comment (eeeeeta) [inline src/types/content.rs]: nit: `whether`

File: src/types/content.rs [hunk 12]
```diff
@@ -0,0 +1,176 @@
+use std::collections::HashMap;
+use ::types::messages::{ImageInfo};
```
Comment (eeeeeta) [inline src/types/content.rs]: is the `::` prefix in this `use` declaration necessary? if not, please remove (I admit, my own understanding of rust's module system may be at fault here)

File: src/types/content.rs [hunk 13]
```diff
@@ -0,0 +1,176 @@
...
+    Public,
+    Invite,
+    // reserved keywords
+    // Knock,
+    // Private,
+}
+
+#[derive(Serialize, Deserialize, Debug)]
+#[serde(rename_all="snake_case")]
+/// Defines what Membership a user in a room
```
Comment (eeeeeta) [inline src/types/content.rs]: grammar nit: "defines what **m**embership a user *has* in a room" perhaps?

File: src/types/content.rs [hunk 14]
```diff
@@ -0,0 +1,176 @@
...
+    #[serde(rename="m.room.create")]
+    Create { creator: String },
+    #[serde(rename="m.room.join_rules")]
+    JoinRule{ join_rule: JoinRule },
+    #[serde(rename="m.room.member")]
+    Member {
+        avatar_url: Option<String>,
+        displayname: Option<String>,
+        membership: Membership,
+        is_direct: 	Option<bool>,
```
Comment (eeeeeta) [inline src/types/content.rs]: nit: random tab

File: src/types/events.rs [hunk 15]
```diff
@@ -0,0 +1,84 @@
...
+    // extra
+    pub age: Option<u64>,
+    pub txn_id: Option<String>,
+    pub redacts: Option<String>,
+    pub membership: Option<String>,
+}
+
+#[derive(Serialize, Deserialize, Debug)]
+#[serde(untagged)]
+/// for now events have to exactly follow the specs
```
Comment (eeeeeta) [inline src/types/events.rs]: I think we could just take this doc comment out

**Review by eeeeeta — CHANGES_REQUESTED** [review]
Thanks *very* much for taking the time to contribute! These changes are really helpful, and increase the capability of the library by a huge amount! (You even added tests, which are absolutely wonderful as well!) I just have some (minor) change requests for you to review, if that's alright, and then I'll gladly merge this.

**Comment by eeeeeta** [comment]
FWIW, if you want to discuss this with me on matrix, feel free to PM me (I realise I may have accidentally declined an invite from you earlier)

File: Cargo.toml (hunk 4 above)
Comment (DigitalChiller) [reply Cargo.toml]: As far as i understand, feature names must be globally unique, so I prefixed them with "gitm".

File: src/types/messages.rs [hunk 16]
```diff
@@ -0,0 +1,242 @@
...
+        #[serde(default)]
+        formatted_body: Option<String>,
+        /// The format of the formatted body (if the message is formatted).
+        #[serde(default)]
+        format: Option<String>
+    },
+    #[serde(rename="m.image")]
+    /// This message represents a single image and an optional thumbnail.
+    /// Information about an image and it's thumbnail.
+    Image{
```
Comment (eeeeeta) [inline src/types/messages.rs]: You can be more DRY here by changing this to `Image(Image)`, since you made a new `Image` struct with these attributes...

**Review by eeeeeta — CHANGES_REQUESTED** [review]

File: src/types/content.rs (hunk 12 above)
Comment (DigitalChiller) [reply src/types/content.rs]: no you don't need it

File: src/types/messages.rs (hunk 16 above)
Comment (DigitalChiller) [reply src/types/messages.rs]: Instantiating Image(Image) is different than instantiating Image{ fields... } and the other message types are of the later format, so I'll remove the image struct.

**Review by eeeeeta — APPROVED** [review]

**Comment by eeeeeta** [comment]
LGTM, thank you for your contribution!

**Pull request merged**